force-app/main/default/classes/JobApplicationStatusHelper.cls

```diff
@@ -0,0 +1,103 @@
+/**
+ * @description       : This helper Class will create Tasks based on the Status of the Job Application.
+ * @author            : Raghava
+ * @group             : Cloud Smiths
+ * @last modified on  : 07-20-2024
+ * @last modified by  : Raghava
+**/
+public with sharing class JobApplicationStatusHelper {
+
+    public static List<Task> tasksToBeCreated = new List<Task>();
+
+    public static void tasksBasedOnStatus(List<Job_Application__c> newJobApplications, Map<Id, Job_Application__c> oldJobApplicationsMap){
+
+        //Looping over all the Job Application records that are passed through After Update Trigger Context
+        for(Job_Application__c jobApp : newJobApplications){
+            //Check if this is an update operation and there is a change in the Status of the Job Application
+            if(oldJobApplicationsMap != null && jobApp.Status__c != oldJobApplicationsMap.get(jobApp.Id).Status__c){
+                handleStatusChange(jobApp.Status__c, jobApp);
+            }
+            //To create tasks when the Job Application is Inserted.
+            else {
+                handleStatusChange(jobApp.Status__c, jobApp);
+            }
```

Comment:
It looks like the if else statement is performing the same operation. If so, is the if else statement needed?

Reply:
@triden92 Can you please help me which two if-else statements are doing the same operation. 

In line 17, 21 - this if-else statement is used to check whether the Trigger context is After Update / After Insert based on verifying if Trigger.oldMap collection is not null. 

Another if statement I used is in Line 26, to check the collection 'tasksToBeCreated' is not empty which obvious will not be but I just added this as a habit. Do you want me remove this?

Reply:
Line 17 says if the oldJobApplicationsMap is not null and the jobApp Status is not equal to the old value, then fire the handleStatusChange method, then Line 20 says if it doesn't meet the above criteria to still trigger the handleStatusChange method. Either way, the same method is being fired.

Reply:
@triden92

It is because to avoid the de-referencing a null pointer value when we tried to call the method - 'tasksBasedOnStatus' for the After Insert context. This method is passed Trigger.new & Trigger.oldMap records as arguments when called. Since for the After Context scenario, the Trigger.oldMap values will be null, it will lead to "WET" code, when we design two separate methods for After Insert & After Update context. And as you pointed out, since anyhow we are calling the same method, one more reason to include this is in the 'After Update' context we only want to fire the method here when there is a change in Status of the Job Application. If we didn't include this if-else, 'handleStatusTasks' will be called when everytime a record is updated.